File: 3.cameraandmatrices/mobile/Engine/Source/Scenes/CCSceneAppUI.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCSceneAppUI.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"


CCSceneAppUI::CCSceneAppUI()
{
    camera = NULL;

    orientationCallback = NULL;
    updatingOrientation = false;

    scrollBar = NULL;
}


void CCSceneAppUI::setup()
{
    if( camera == NULL )
    {
        camera = new CCCameraAppUI();
        gEngine->addCamera( camera );
        camera->setupViewport( 0.0f, 0.0f, 1.0f, 1.0f );
    }
    camera->setCameraWidth( 100.0f );

    refreshCameraView();
    lockCameraView();
}


void CCSceneAppUI::destruct()
{
    orientationCallback.deleteObjectsAndList();

    if( camera != NULL )
    {
        gEngine->removeCamera( camera );
        delete camera;
    }
    
    super::destruct();
}


// CCSceneBase
void CCSceneAppUI::updateScene(const CCTime &gameTime)
{
    super::updateScene( gameTime );
}


void CCSceneAppUI::updateCamera(const CCTime &gameTime)
{
    const float lookAtSpeed = 1.5f;
	if( camera->interpolateCamera( gameTime.delta, lookAtSpeed ) )
    {
	}
    else
    {
        if( cameraScrolling )
        {
            cameraScrolling = false;
            lockCameraView();
        }
        
        if( updatingOrientation )
        {
            updatingOrientation = false;
            refreshCameraView();
            lockCameraView();
        }
    }
}


const bool CCSceneAppUI::render(const CCCameraBase *inCamera, const int pass, const bool alpha)
{
    if( camera == inCamera )
    {
        renderObjects( pass, alpha );
        return true;
    }
    return false;
}


void CCSceneAppUI::renderOctreeObject(CCSceneObject *object, const CCCameraBase *inCamera, const int pass, const bool alpha)
{
    if( camera == inCamera )
	{
        object->render( alpha );
	}
}


void CCSceneAppUI::render2DBackground(const uint inCameraIndex)
{
    if( inCameraIndex == camera->getIndex() )
    {
        for( int i=0; i<childScenes.length; ++i )
        {
            childScenes.list[i]->render2DBackground( inCameraIndex );
        }
    }
}


void CCSceneAppUI::render2DForeground(const uint inCameraIndex)
{
    if( inCameraIndex == camera->getIndex() )
    {
        for( int i=0; i<childScenes.length; ++i )
        {
            childScenes.list[i]->render2DForeground( inCameraIndex );
        }
    }
}


void CCSceneAppUI::beginOrientationUpdate()
{
    camera->setCameraWidth( camera->cameraWidth );

    for( int i=0; i<orientationCallback.length; ++i )
    {
        orientationCallback.list[i]->run();
    }
    
    updatingOrientation = true;
}


void CCSceneAppUI::finishOrientationUpdate()
{
    updatingOrientation = true;
}


void CCSceneAppUI::scrollCameraToTop()
{
    camera->targetLookAt.y = sceneTop;
    camera->flagUpdate();
}


void CCSceneAppUI::refreshCameraView()
{
    sceneLeft = 0.0f;
    sceneRight = 0.0f;
    sceneTop = 0.0f;
    sceneBottom = 0.0f;
}


// Lock the camera on something interesting
void CCSceneAppUI::lockCameraView()
{
    if( cameraScrolling )
    {
        return;
    }
    
    camera->flagUpdate();
    camera->targetLookAt.x = 0.0f;
    camera->targetLookAt.y = 0.0f;
}

File: 3.cameraandmatrices/webgl/source/js/engine/scenes/CCSceneAppUI.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCSceneAppUI.js
 * Description : AppUI scene template.
 *
 * Created     : 20/10/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCSceneAppUI()
{
    this.construct();
}
CopyPrototype( CCSceneAppUI, CCSceneBase, "CCSceneBase" );


CCSceneAppUI.prototype.construct = function()
{
	this.CCSceneBase_construct();

	this.camera = false;
	
	this.resizeCallbacks = new Array();
	this.resizing = false;
    
    this.scrollBar = false;
}


CCSceneAppUI.prototype.setup = function()
{
	if( this.camera == false )
	{
		var camera = this.camera = new CCCameraAppUI();
		gEngine.addCamera( camera );
		camera.setupViewport( 0.0, 0.0, 1.0, 1.0 );
		camera.setCameraWidth( 100.0 );
	}
	
    this.refreshCameraView();
    this.lockCameraView();
}


CCSceneAppUI.prototype.resize = function()
{   
	this.camera.setCameraWidth( 100.0 );

	var resizeCallbacks = this.resizeCallbacks;
	var length = this.resizeCallbacks.length;
    for( var i=0; i<length; ++i )
    {
        resizeCallbacks[i].run();
    }
    
    this.resizing = true;
}


CCSceneAppUI.prototype.update = function(delta)
{
	this.CCSceneBase_update( delta );
}


CCSceneAppUI.prototype.updateCamera = function(delta)
{
	var lookAtSpeed = 1.5;
	if( this.camera.interpolateLookAt( delta, lookAtSpeed ) )
    {
        // Tell the scroll bar where to go
        if( this.scrollBar != false )
        {
            //this.scrollBar->reposition( camera->getLookAt().y, cameraWidth, cameraHeight );
        }
	}
    else
    {
        if( this.cameraScrolling )
        {
            this.cameraScrolling = false;
            this.lockCameraView();
        }
        
        if( this.resizing )
        {
            this.resizing = false;
            this.refreshCameraView();
            this.lockCameraView();
        }
    }
}


CCSceneAppUI.prototype.render = function(renderer, camera, pass, alpha)
{
	if( this.camera == camera )
	{
		this.CCSceneBase_render( renderer, camera, pass, alpha );
	}
}


CCSceneAppUI.prototype.scrollCameraToTop = function()
{
    this.camera.targetLookAt[1] = this.sceneTopY;
    //camera->flagUpdate();
}


CCSceneAppUI.prototype.refreshCameraView = function()
{
    this.sceneTopY = 0.0;
    this.sceneBottomY = 0.0;
}


// Lock the camera on something interesting
CCSceneAppUI.prototype.lockCameraView = function()
{
    if( this.cameraScrolling )
    {
        return;
    }
    
    //camera->flagUpdate();
    gEngine.renderer.pendingRender = true;
    this.camera.targetLookAt[0] = 0.0;
    this.camera.targetLookAt[1] = 0.0;
}